providers/user.rb

```diff
@@ -0,0 +1,87 @@
+use_inline_resources if defined?(use_inline_resources)
+
+# Support whyrun
+def whyrun_supported?
+  true
+end
+
+action :upsert do
+  client = get_client.use(@new_resource.database)
+
+  if @current_resource.exists
+    converge_by("Update #{ @new_resource }") do
+      Chef::Log.info "#{ @new_resource } already exists in database #{ @new_resource.database } - updating."
+      client.database.users.update(@new_resource.username,
+                                   :password => @new_resource.password,
+                                   :roles => @new_resource.roles,
+                                   :database => @new_resource.database)
+    end
+  else
+    converge_by("Create #{ @new_resource }") do
+      Chef::Log.info "Creating #{ @new_resource } in database #{ @new_resource.database }"
+      client.database.users.create(@new_resource.username,
+                                   :password => @new_resource.password,
+                                   :roles => @new_resource.roles,
+                                   :database => @new_resource.database)
+
+    end
+  end
+end
+
+action :delete do
+  client = get_client.use(@new_resource.database)
+
+  if @current_resource.exists
+    converge_by("Delete #{ @new_resource }") do
+      Chef::Log.info "#{ @new_resource } deleting."
+      client.database.users.delete(@new_resource.username)
+    end
+  else
+    Chef::Log.info "#{ @new_resource } does not exists."
+  end
+end
+
+def get_client
+  require 'rubygems'
+  require 'mongo'
+
+  config = node['mongodb3']['config']['mongod']
+  client = Mongo::Client.new(
+      'mongodb://%s:%s' % [config['net']['bindIp'] || '127.0.0.1', config['net']['port']],
+      :database => 'admin',
+      :connect => :direct
+  )
+
+  if config['security']['authorization'] == 'enabled'
+    admin = node['mongodb3']['admin']
+
+    if admin['username'] and admin['password']
+      client = client.with(user: admin['username'], password: admin['password'])
```

Comment:
Let's not to create attribute dependency here. having admin pwd in attribute will be the problem in security since node file can be downloaded to someones local machine.